// @cloudflare/types/es/api/alert.js
import { eg } from '@cloudflare/util-en-garde';
// NotificationMechanism is specifies how a notification is delivered -
// via email, SMS, PagerDuty, etc
// type INotificationMechanism = 'email' | 'sms' | 'pagerduty';

export var AlertIntegration = eg.object({
  id: eg.string,
  name: eg.string,
  createdAt: eg.string
});

/**
 * A connection request to Pager Duty that is created before we attempt to link
 * PagerDuty to this account
 */

var PagerDutyIntegrationRequest = eg.object({
  id: eg.string
});
var AlertWebhookType = eg.union([eg.literal('slack'), eg.literal('gchat'), eg.literal('discord'), eg.literal('generic'), eg.literal('splunk'), eg.literal('datadog'), eg.literal('opsgenie'), eg.literal('feishu')]);
var AlertWebhookIntegration = eg.object({
  id: eg.string,
  name: eg.string,
  url: eg.string,
  createdAt: eg.string,
  secret: eg.string.optional,
  last_success: eg.string.optional,
  last_failure: eg.string.optional
});
export var AlertEligibleIntegrations = eg.record(eg.string, eg.object({
  eligible: eg.boolean,
  ready: eg.boolean,
  enabled: eg.boolean,
  type: eg.string
}));
export var EmailParameters = eg.object({
  recipients: eg.array(eg.string)
});
// ConditionFilter is pretty much just a condition.
var ConditionFilter = eg.object({
  '>=': eg.array(eg.any)
});
// Type AlertPolicy is the type that represents an alert policy
export var AlertPolicy = eg.object({
  id: eg.string.optional,
  name: eg.string,
  description: eg.string.optional,
  enabled: eg.boolean,
  emails: EmailParameters,
  pager_duty: eg.array(eg.string).optional,
  webhooks: eg.array(eg.string).optional,
  alert_type: eg.string,
  conditions: ConditionFilter.optional,
  filters: eg.record(eg.string, eg.union([eg.array(eg.string), eg.null])).optional
});
export var AlertDeliveryMechanism = eg.object({
  id: eg.string,
  name: eg.string.optional
});
var AlertMechanismMap = eg.record(eg.string, eg.array(AlertDeliveryMechanism));
export var AlertPolicyAPIV2 = eg.object({
  id: eg.string.optional,
  // UUID, nullable for creating new API's
  name: eg.string,
  description: eg.string.optional,
  enabled: eg.boolean,
  conditions: ConditionFilter.optional,
  mechanisms: AlertMechanismMap,
  filters: eg.record(eg.string, eg.union([eg.array(eg.string), eg.null])).optional,
  alert_type: eg.string
});
export var AvailableAlert = eg.object({
  display_name: eg.string,
  type: eg.string,
  description: eg.string.optional
});
export var TestAlertRequest = eg.object({});
export var StatuspageComponent = eg.object({
  id: eg.string,
  name: eg.string,
  parent_component_id: eg.string.optional
});
